architecture/2024-08-applicationdeployment-controller.md

```diff
@@ -0,0 +1,370 @@
+- `spec.parameters`: The Bicep parameters file that defines the parameters for the template
+- `status`: The status of the deployment, which will be one of the following values:
+  - `Pending`: The deployment is pending
+  - `Deployed`: The deployment has been completed
+  - `Missing Dependencies`: The deployment is missing dependencies
+
+#### `rad applicationdeployment generate` Command
+
+The `rad applicationdeployment generate` command will be a new command that will generate an ApplicationDeployment resource from a Bicep template and parameters file. The command will take the following arguments:
+
+- `-t, --template`: The path to the Bicep template file
+- `-p, --parameters`: The path to the Bicep parameters file
+- `-o, --output`: The path to the output file where the ApplicationDeployment resource will be written
+
+The command will read the Bicep template and parameters files and generate an ApplicationDeployment resource with the template and parameters set. The command will write the ApplicationDeployment resource to the output file.
+
+#### ApplicationDeployment Controller
+
+The `radius-controller` will be updated to include a new controller that watches for ApplicationDeployment resources. When an ApplicationDeployment resource is created or updated, the controller will read the Bicep template and parameters from the ApplicationDeployment resource and send them to the `radius-bicep` service. The `radius-bicep` service will compile the Bicep and bicepparam files into ARM JSON and return the ARM JSON to the `radius-controller`.
+
+For each of the resources in the template, the `radius-controller` will determine if the dependencies for the resource already exist by querying UCP. If they do not, then the controller will mark the ApplicationDeployment resource as "Missing Dependencies" with the resource that is missing and will not deploy the resource. If the dependencies do already exist for each of the resources in the template, then the controller will send the ARM JSON payload to UCP to deploy the resources. Once the deployment is complete, The controller will trigger a re-deployment of resources with status "Missing Dependencies" that depend on the newly deployed resources. The controller will update the status of the ApplicationDeployment resource to "Deployed" once all resources have been deployed.
+
+#### Bicep Service
+
+The `radius-bicep` service will be a new service that is responsible for compiling ARM JSON from Bicep manifests. The service will expose an HTTP endpoint that accepts a payload containing a Bicep template and bicepparam file and returns the compiled ARM JSON. The service will use the Bicep compiler to compile the Bicep manifest and will return the ARM JSON to the caller. The service will be stateless and will not store any data.
+
+### Alternatives Considered
+
+We could package the Bicep compiler into the `radius-controller` service instead of creating a separate service.
```

Comment:
I think that service should be separate since it could be used by some other components in future.